exporter/opensearchexporter/trace_bulk_indexer.go

```diff
@@ -0,0 +1,245 @@
+				now := time.Now()
+				ssoEvent.ObservedTimestamp = &now
+			}
+		}
+	}
+
+	dataStream := DataStream{}
+	if tbi.dataset != "" {
+		dataStream.Dataset = tbi.dataset
+	}
+
+	if tbi.namespace != "" {
+		dataStream.Namespace = tbi.namespace
+	}
+
+	if dataStream != (DataStream{}) {
+		dataStream.Type = "span"
+		sso.Attributes["data_stream"] = dataStream
+	}
+
+	sso.InstrumentationScope.Name = scope.Name()
+	sso.InstrumentationScope.DroppedAttributesCount = scope.DroppedAttributesCount()
+	sso.InstrumentationScope.Version = scope.Version()
+	sso.InstrumentationScope.SchemaURL = schemaURL
+	sso.InstrumentationScope.Attributes = scope.Attributes().AsRaw()
+
+	if span.Links().Len() > 0 {
+		sso.Links = make([]SSOSpanLinks, span.Links().Len())
+		for i := 0; i < span.Links().Len(); i++ {
+			link := span.Links().At(i)
+			ssoLink := &sso.Links[i]
+			ssoLink.Attributes = link.Attributes().AsRaw()
+			ssoLink.DroppedAttributesCount = link.DroppedAttributesCount()
+			ssoLink.TraceID = link.TraceID().String()
+			ssoLink.TraceState = link.TraceState().AsRaw()
+			ssoLink.SpanID = link.SpanID().String()
+		}
+	}
+	return json.Marshal(sso)
+}
+
+func (tbi *traceBulkIndexer) processItemFailure(resp opensearchutil.BulkIndexerResponseItem, itemErr error, traces ptrace.Traces) {
+	switch {
+	case shouldRetryEvent(resp.Status):
+		// Recoverable OpenSearch error
+		tbi.appendRetryTraceError(responseAsError(resp), traces)
+	case resp.Status != 0 && itemErr == nil:
+		// Non-recoverable OpenSearch error while indexing document
+		tbi.appendPermanentError(responseAsError(resp))
+	default:
+		// Encoding error. We didn't even attempt to send the event
+		tbi.appendPermanentError(itemErr)
+	}
+}
+
+// responseAsError converts an opensearchutil.BulkIndexerResponseItem.Error into an error
+func responseAsError(item opensearchutil.BulkIndexerResponseItem) error {
+	errorJSON, _ := json.Marshal(item.Error)
+	return errors.New(string(errorJSON))
+}
+
+func shouldRetryEvent(status int) bool {
+	var retryOnStatus = []int{500, 502, 503, 504, 429}
+	for _, retryable := range retryOnStatus {
+		if status == retryable {
+			return true
+		}
+	}
+	return false
+}
+
+func (tbi *traceBulkIndexer) newBulkIndexerItem(document []byte) opensearchutil.BulkIndexerItem {
+	body := bytes.NewReader(document)
+	item := opensearchutil.BulkIndexerItem{Action: "create", Index: tbi.getIndexName(), Body: body}
+	return item
+}
+
+func (tbi *traceBulkIndexer) getIndexName() string {
+	return strings.Join([]string{"ss4o_traces", tbi.dataset, tbi.namespace}, "-")
+}
+
+func newOpenSearchBulkIndexer(client *opensearch.Client, onIndexerError func(context.Context, error)) (opensearchutil.BulkIndexer, error) {
+	return opensearchutil.NewBulkIndexer(opensearchutil.BulkIndexerConfig{
+		NumWorkers: 1,
+		Client:     client,
+		OnError:    onIndexerError,
+	})
+}
+
+func forEachSpan(td ptrace.Traces, iterator func(pcommon.Resource, string, pcommon.InstrumentationScope, string, ptrace.Span)) {
```

Comment:
```suggestion
func forEachSpan(td ptrace.Traces, visitor func(pcommon.Resource, string, pcommon.InstrumentationScope, string, ptrace.Span)) {
```

Since this is invoking the provided function on each item in the collection it is a visitor rather than an iterator, which would return the next item in the collection on each invocation.